### PR TITLE
Added connection pool class

### LINKED ISSUE #39: add connection pooling
Can add connection pooling？like http://pub.dartlang.org/packages/postgresql

### PR DESCRIPTION
**Pull request by thosakwe**
Should resolve #39.

cc @vadimtsushko @Pacane I'd love to hear your opinions before merging.

### CODE REVIEW

File: lib/src/connection_pool.dart [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+///
+/// The connection pool lazily connects to the database; that is to say, it only opens as many
+/// connections as it needs to. If it is only ever called once, then it will only ever connect once.
+class ConnectionPool {
+  List<Db> _connections = [];
+  int _index = 0;
+  Pool _pool;
+
+  /// The maximum number of concurrent connections allowed.
+  final int concurrency;
```
Comment (Pacane) [inline lib/src/connection_pool.dart]: I'd remove that comment and rename the variable to match what's in the comment. e.g. `maxNumberOfConcurrentConnections`

File: lib/src/connection_pool.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  /// A [DbFactory], a parameterless function that returns a [Db]. The function can be asynchronous if necessary.
+  final DbFactory dbFactory;
```
Comment (Pacane) [inline lib/src/connection_pool.dart]: is there a reason why this is public?

File: lib/src/connection_pool.dart [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+/// This is used in the [ConnectionPool] class to connect to a database on-the-fly.
+typedef FutureOr<Db> DbFactory();
+
+/// A connection pool that limits the number of concurrent connections to a MongoDB server.
+///
+/// The connection pool lazily connects to the database; that is to say, it only opens as many
+/// connections as it needs to. If it is only ever called once, then it will only ever connect once.
+class ConnectionPool {
+  List<Db> _connections = [];
+  int _index = 0;
```
Comment (Pacane) [inline lib/src/connection_pool.dart]: rename this more clearly.

Comment (Pacane) [inline lib/src/connection_pool.dart]: e.g.: `currentAmountOfConnections`

File: lib/src/connection_pool.dart [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        await db.open();
+        _connections.add(db);
+        return db;
+      }
+    });
+  }
+
+  /// Closes all active database connections.
+  Future close() {
+    return Future
```
Comment (Pacane) [inline lib/src/connection_pool.dart]: Shouldn't close() decrement the index or something like that?

**Comment by Pacane** [comment]
Also, I just remembered, that I always used https://github.com/redstone-dart/connection_pool when I wanted a connection pool. We might as well use that one, I don't know.

**Comment by vadimtsushko** [comment]
I would rather prefer not to add yet another external dependency to repositary

**Comment by vadimtsushko** [comment]
I did not know much about connection pooling, but I do not see how resources meant to be deallocated here. It looks like client get simple `Db` object.  What if first client invoke `close` on that `Db`? It looks like fourth client will get `Db` in invalid state? Looking on https://github.com/xxgreg/dart_postgresql/blob/master/lib/pool.dart it looks like they trying to check for many possible bad scenarios and make some remedies.

**Comment by vadimtsushko** [comment]
I've never use something like connection pooling myself and do not now what properties are essential for these solutions. But I'm afraid that with pooling clients expect some sort of additional features. So I would rather not have pooling at all until these requirements could not be met.

**Comment by thosakwe** [comment]
Is there a way to check if a `Db` is closed, then?

File: lib/src/connection_pool.dart (hunk 3 above)
Comment (thosakwe) [inline lib/src/connection_pool.dart]: 👍

File: lib/src/connection_pool.dart (hunk 1 above)
Comment (thosakwe) [inline lib/src/connection_pool.dart]: 👍

File: lib/src/connection_pool.dart (hunk 2 above)
Comment (thosakwe) [inline lib/src/connection_pool.dart]: None in particular. I can mark this as private, and then expand on the comment in the constructor to make it clearer.

File: lib/src/connection_pool.dart (hunk 4 above)
Comment (thosakwe) [inline lib/src/connection_pool.dart]: No need, because `close()` will close all the connections in the pool.

**Comment by thosakwe** [comment]
Also, I'm not much in favor of adding an external dependency; however, I made an exception because this is developed by the Dart Team, as Vadim said.

**Comment by thosakwe** [comment]
Lastly, in regard to formatting: I'm pretty sure there is a way to ensure that new pulls are properly formatted, and fail builds if they do not match the output of `dartfmt`.

**Comment by vadimtsushko** [comment]
If i recall correctly there was some private member in Db. Like _closed. Which was set to true on close method.

**Comment by thosakwe** [comment]
I think I will merge it, but first I must complete the suggestions Joel made.

**Comment by thosakwe** [comment]
Submitting tomorrow, woohoo!

**Comment by thosakwe** [comment]
Apologies for not following through and submitting this. Once I get into the swing of it, however, the PR's will be coming in pretty close succession.

**Pull request merged**